### PR TITLE
fix task4

### PR DESCRIPTION
**Pull request by lizasmirnova22**
добавлена генерация параметров для задания 4 весна

### CODE REVIEW

File: src/prog_questions/spring/QuestionN4.py [hunk 1]
```diff
@@ -1,87 +1,144 @@
 from ..QuestionBase import QuestionBase, Result
-from ..utility import CppProgramRunner, CompilationError, ExecutionError, InternalError
+from ..utility import CppProgramRunner, CompilationError, ExecutionError
+from dataclasses import dataclass, field
```
Comment (Anna-Moiseeva-3341) [inline src/prog_questions/spring/QuestionN4.py]: `field` импортируется, при этом в программе не используется

File: src/prog_questions/spring/QuestionN4.py [hunk 2]
```diff
@@ -100,282 +157,149 @@ def _get_example_table(self) -> str:
...
+    <li>Слова внутри группы отсортированы лексикографически.</li>
+</ul>
+<p><i>Если после фильтрации не осталось ни одной группы, выведите <code>NO_DATA</code>.</i></p>"""
 
-int main() {
-    // Ваш код
-    return 0;
-}
-"""
+    # Пример
```
Comment (Anna-Moiseeva-3341) [inline src/prog_questions/spring/QuestionN4.py:192]: В 4.1 пример захардкожен без учёта фильтра.
Например, если сгенерирован фильтр exact_count = 2, то группа c: cat из примера должна быть убрана, потому что там одно слово.

File: src/prog_questions/spring/QuestionN4.py [hunk 3]
```diff
@@ -100,282 +157,149 @@ def _get_example_table(self) -> str:
...
-    <tbody>
-        <tr class="r0 lastrow">
-            <td class="cell c0"><pre class="tablecell">{inp}</pre></td>
-            <td class="cell c1"><pre class="tablecell">{out}</pre></td>
-            <td class="cell c2 lastcol">{note}</td>
-        </tr>
-    </tbody>
-</table>'''
+    # Пример
+    example_inp = "3 3 3 5 5 0"
```
Comment (Anna-Moiseeva-3341) [inline src/prog_questions/spring/QuestionN4.py]: В 4.2 пример тоже захардкожен. Он корректно работает только для операции shift и подходящей маски.

File: src/prog_questions/spring/QuestionN4.py [hunk 4]
```diff
@@ -385,49 +309,135 @@ def preloaded_code(self) -> str:
...
             raise
-        except Exception as e:
-            return Result.Fail("INTERNAL", "No error expected", f"Internal error: {str(e)}")
+
+        min_tests, max_tests = 5, 25
+        num_tests = max(int(min_tests + self.strictness * (max_tests - min_tests)), 8)
+        edge_cases = 4  # количество краевых тестов
+
+        for test_num in range(num_tests):
+            generator = Task4ConfigGenerator(self.seed + test_num, self.strictness)
```
Comment (Anna-Moiseeva-3341) [inline src/prog_questions/spring/QuestionN4.py:455]: Текст задания строится из self.config, но в test() конфиг генерируется заново. Из-за этого сохраняется только тип без конкретных параметров.

File: src/prog_questions/spring/QuestionN4.py [hunk 5]
```diff
@@ -385,49 +309,135 @@ def preloaded_code(self) -> str:
...
+
+            try:
+                output = runner.run(input_data).replace('\r', '').strip()
+            except ExecutionError as e:
+                return Result.Fail(input_data, expected, str(e))
+
+            if output != expected:
+                return Result.Fail(input_data, expected, output)
+
+        return Result.Ok()
```
Comment (aassiya) [inline src/prog_questions/spring/QuestionN4.py]: Нет пустой строки в конце файла

File: tests/QuestionN4_spring_test.py [hunk 6]
```diff
@@ -145,4 +158,73 @@ def test_all_task_types_possible(self):
...
+            else:
+                assert 'Динамический учёт слов' in q.questionText
+
+    def test_without_override_generates_random_types(self):
+        """Без параметра task_type_override типы должны меняться при разных seed"""
+        types_seen = set()
+        for seed in range(20):
+            q = QuestionN4(seed=seed)
+            types_seen.add(q.task_type)
+        assert len(types_seen) == 3, f"Ожидались все 3 типа, получены {types_seen}"
```
Comment (aassiya) [inline tests/QuestionN4_spring_test.py]: Нет пустой строки в конце файла

File: src/prog_questions/spring/QuestionN4.py [hunk 7]
```diff
@@ -1,87 +1,144 @@
...
 import random
-from textwrap import dedent
-import subprocess
-import os
-import sys
 from collections import defaultdict, Counter
+from textwrap import dedent
+
+
+
```
Comment (aassiya) [inline src/prog_questions/spring/QuestionN4.py]: Лишняя пустая строка

File: src/prog_questions/spring/QuestionN4.py [hunk 8]
```diff
@@ -1,87 +1,144 @@
...
+        }
+        return names.get(self.task_type, self.task_type)
+
+
+class Task4ConfigGenerator:
+    """Генератор конфигураций для задания 4."""
+    # Константы для генерации
+    WORDS_POOL = ['apple', 'ant', 'apricot', 'banana', 'bat', 'bear',
+                  'cat', 'car', 'cup', 'dog', 'duck', 'deer', 'elephant']
+    VALUE_RANGE = (1, 200)
```
Comment (aassiya) [inline src/prog_questions/spring/QuestionN4.py:46]: VALUE_RANGE объявлен, но нигде не используется. Вместо него используется: `random.randint(1, 200)`

File: src/prog_questions/spring/QuestionN4.py [hunk 9]
```diff
@@ -385,49 +309,135 @@ def preloaded_code(self) -> str:
...
+                config.words = ['cat', 'car', 'cup']
+        elif config.task_type == '4.2':
+            if test_num == 0:
+                config.numbers = []
+            elif test_num == 1:
+                config.numbers = [1, 2, 3]
+            elif test_num == 2:
+                config.numbers = [5, 5, 0] if (5 & config.mask) == config.mask else [1, 1]
+            elif test_num == 3:
+                config.numbers = [config.mask] * (config.min_freq + 2) + [0]
```
Comment (aassiya) [inline src/prog_questions/spring/QuestionN4.py]: Для 4.2 `test_num == 2` и `test_num == 3` в список numbers добавляется 0. Но по условию задачи 4.2 ноль - это терминатор входного потока, он не должен участвовать в подсчёте частоты. При этом `_generate_input` добавляет ещё один 0 в конец строки.

File: tests/QuestionN4_spring_test.py
Comment (aassiya) [inline tests/QuestionN4_spring_test.py:1]: Возможно стоит добавить отдельные тесты на edge-cases
